### ISSUE TITLE
Tunneled SSH connection fails because of non-resolved host name

### ISSUE THREAD

**Issue by TheJJ** [ronf/asyncssh#229]
I'm trying to do a SSH connection within an existing SSH connection to achieve `ssh -J jumphost destinationhost` with `asyncssh`.

``` python
JUMPHOST = 'jump.host.domain'
TARGETHOST = 'funny.switch.domain'

async def lol():
    async with asyncssh.connect(JUMPHOST) as tunnel:
        async with asyncssh.connect(host, tunnel=tunnel, username=xxx, password=xxx) as conn:
            return await conn.run("version")
```

This fails:
```
Traceback (most recent call last):
  File "./zonegen.py", line 110, in <module>
    main()
  File "./zonegen.py", line 69, in main
    loop.run_until_complete(lol())
  File "./zonegen.py", line 39, in lol
    password=xxx) as conn:
  File "/usr/lib/python3.7/site-packages/asyncssh/misc.py", line 192, in __aenter__
    self._result = yield from self._coro
  File "/usr/lib/python3.7/site-packages/asyncssh/connection.py", line 5755, in connect
    'Opening SSH connection to'))
  File "/usr/lib/python3.7/site-packages/asyncssh/connection.py", line 167, in _connect
    yield from conn.wait_established()
  File "/usr/lib/python3.7/site-packages/asyncssh/connection.py", line 1949, in wait_established
    yield from self._waiter
  File "/usr/lib/python3.7/site-packages/asyncssh/connection.py", line 722, in connection_made
    self._connection_made()
  File "/usr/lib/python3.7/site-packages/asyncssh/connection.py", line 2452, in _connection_made
    self._peer_addr, port)
  File "/usr/lib/python3.7/site-packages/asyncssh/connection.py", line 607, in _match_known_hosts
    match_known_hosts(known_hosts, host, addr, port)
  File "/usr/lib/python3.7/site-packages/asyncssh/known_hosts.py", line 324, in match_known_hosts
    known_hosts = known_hosts.match(host, addr, port)
  File "/usr/lib/python3.7/site-packages/asyncssh/known_hosts.py", line 236, in match
    x509_subjects, revoked_subjects = self._match(host, addr, port)
  File "/usr/lib/python3.7/site-packages/asyncssh/known_hosts.py", line 172, in _match
    ip = ip_address(addr) if addr else None
  File "/usr/lib/python3.7/site-packages/asyncssh/misc.py", line 140, in ip_address
    return ipaddress.ip_address(_normalize_scoped_ip(addr))
  File "/usr/lib/python3.7/ipaddress.py", line 54, in ip_address
    address)
ValueError: 'funny.switch.domain' does not appear to be an IPv4 or IPv6 address
```

This seems to be because `connection.SSHConnection.connection_made` calls to `transport.get_extra_info('peername')`, which returns an IP address for a real `socket`, but as the connection is tunneled, `transport` is a `asyncssh.channel.SSHTCPChannel`. `get_extra_info('peername')` for the `SSHTCPChannel` doesn't return the IP address, instead it returns the hostname. The known-hosts-checker expects an IP-address and thus fails.

Am I doing this wrong? Is this the right way to do `ssh -J jumphost destination`? Or is this a bug?

**Comment by ronf** [comment]
When using a jump host, the IP address of the final target (funny.switch.domain in this example) isn't resolved on the client machine calling into AsyncSSH. The hostname string is passed to the jump host and it would do the DNS resolution, without providing back IP address information to the client. That's why the peer information is returning a hostname instead of an IP address.

However, the fact that this hostname is being passed into the known_hosts code as an address is definitely a bug in AsyncSSH. It should probably be passing in `None` there, with an intent that only known host entries matching the host by name would be matched. I suspect that this error was introduced when I recently added the reverse-direction connection support, but I'll need to do a bit of digging about that and look over the code a bit more closely to determine the best way to correct this.

In the meantime, you can reverse back to AsyncSSH 1.16.1 to avoid this problem. I'll try to get a fix into the "develop" branch in the next few days. Thanks for the report!

**Comment by ronf** [comment]
I've just checked commit 7f1db95 into the "develop" branch to address a few different problems related to the setting of host name and peer address when tunneled SSH connections are involved. I believe it should fix the issue you reported above. I've also made the code better handle CIDR-style matching in the case where a numeric IP address is used as a tunnel destination.

While investigating this, I also discovered that the helper methods connect_ssh() and connect_reverse_ssh() were previously not properly marked as async context managers, and so I fixed that in commit 53642fd. This should now let you use syntax like:

```python
    async with asyncssh.connect(JUMPHOST) as tunnel:
        async with tunnel.connect_ssh(host, username=xxx, password=xxx) as conn:
            return await conn.run("version")
```

If you still see any issues with this, please let me know. Thanks very much for the report, and for your analysis. It definitely helped me get to the bottom of the issue quickly.

**Comment by ronf** [comment]
These fixes are now available in the AsyncSSH 1.18.0 release!
